p2p/src/authenticated/actors/tracker/actor.rs

```diff
@@ -204,189 +170,145 @@ impl<E: Spawner + Rng + Clock + GClock + Metrics, C: Scheme> Actor<E, C> {
+            for set in self.sets.values_mut() {
+                set.found(&public_key);
             }
+            debug!(peer = ?public_key, "updated peer record");
         }
 
         Ok(())
     }
 
-    fn handle_bit_vec(
-        &mut self,
-        bit_vec: types::BitVec,
-    ) -> Result<Option<Vec<types::PeerInfo<C>>>, Error> {
+    /// Handle an incoming bit vector from a peer.
+    fn handle_bit_vec(&mut self, bit_vec: types::BitVec) -> Result<Vec<types::PeerInfo<C>>, Error> {
         // Ensure we have the peerset requested
-        let set = match self.sets.get(&bit_vec.index) {
-            Some(set) => set,
-            None => {
-                debug!(index = bit_vec.index, "requested peer set not found");
-                return Ok(None);
-            }
+        let Some(set) = self.sets.get(&bit_vec.index) else {
+            // Don't consider unknown indices as errors, just ignore them.
+            debug!(index = bit_vec.index, "requested peer set not found");
+            return Ok(vec![]);
         };
 
         // Ensure that the bit vector is the same size as the peer set
-        if bit_vec.bits.len() != set.order.len() {
+        if bit_vec.bits.len() != set.sorted.len() {
             return Err(Error::BitVecLengthMismatch(
-                set.order.len(),
+                set.sorted.len(),
                 bit_vec.bits.len(),
             ));
         }
 
         // Compile peers to send
-        let mut peers = Vec::new();
-        for (order, bit) in bit_vec.bits.iter().enumerate() {
-            // Check if we have exhausted our known peers
-            let peer = match set.sorted.get(order) {
-                Some(peer) => peer,
-                None => {
-                    if bit {
-                        return Err(Error::BitVecExtraBit);
-                    }
-
-                    // It is ok if the bit vector is smaller than the peer set but
-                    // there should never be bits in these positions.
-                    continue;
-                }
-            };
-
-            // Check if the peer already knows of this peer
-            if bit {
-                continue;
-            }
-
-            // Add the peer to the list if its address is known
-            if *peer == self.crypto.public_key() {
-                peers.push(self.ip_signature.clone());
-                continue;
-            }
-            let Some(peer_info) = self.peers.get(peer).and_then(|r| r.get_peer_info()) else {
-                debug!(?peer, "peer address not known");
-                continue;
-            };
-            peers.push(peer_info.clone());
-        }
-
-        // Return None if no peers to send
-        if peers.is_empty() {
-            return Ok(None);
-        }
+        let mut peers: Vec<_> = bit_vec
+            .bits
+            .iter()
+            .enumerate()
+            .filter(|(_, bit)| !bit) // Only consider peers that are not known to the requester
```

Comment:
nit: `not needed/wanted` (I think `known` may now be too strong of a word for this)